server/recurring/cluster.go

```diff
@@ -0,0 +1,271 @@
+	value, ok := w.wc[word]
+	if !ok {
+		w.wc[word] = [2]int{
+			w.index,
+			1,
+		}
+		w.index++
+		return
+	}
+
+	value[1]++
+	w.wc[word] = value
+}
+
+func (w *WordCount) Iterate(callback func(word string, index int, count int)) {
+	for word, value := range w.wc {
+		callback(word, value[0], value[1])
+	}
+}
+
+func (w *WordCount) Index(word string) int {
+	value, ok := w.wc[word]
+	if !ok {
+		return -1
+	}
+	return value[0]
+}
+
+func (w *WordCount) Minify() map[string]int {
+	index := 0
+	result := make(map[string]int)
+	for word, value := range w.wc {
+		if value[1] == 1 {
+			continue
+		}
+		result[word] = index
+		index++
+	}
+
+	return result
+}
+
+type PreProcessor struct {
+	documents []Document
+	// Word count
+	wc  *WordCount
+	idf map[string]float64
+}
+
+func (p *PreProcessor) AddTransaction(txn *Transaction) {
+	words := cleanStringRegex.FindAllString(txn.OriginalName, len(txn.OriginalName))
+	if txn.OriginalMerchantName != nil {
+		words = append(words, strings.Fields(*txn.OriginalMerchantName)...)
+	}
+
+	wordCounts := make(map[string]int, len(words))
+	// Get the term frequency from the transaction name
+	for _, word := range words {
+		word = strings.ToLower(word)
+		wordCounts[word]++
+		p.wc.Increment(word)
+	}
+
+	tf := make(map[string]float64, len(wordCounts))
+	for word, count := range wordCounts {
+		tf[word] = float64(count) / float64(len(words))
+	}
+
+	p.documents = append(p.documents, Document{
+		Transaction: txn,
+		TF:          tf,
+		TFIDF:       map[string]float64{},
+	})
+}
+
+// PostPrepareCalculations should be called after all transactions have been added to the dataset.
+func (p *PreProcessor) PostPrepareCalculations() {
+	docCount := float64(len(p.documents))
+	p.wc.Iterate(func(word string, _ int, count int) {
+		p.idf[word] = math.Log(docCount / (float64(count) + 1))
+	})
+	var min, max float64
+	for i := range p.documents {
+		document := p.documents[i]
+		for word, tfValue := range document.TF {
+			document.TFIDF[word] = tfValue * p.idf[word]
+		}
+		p.documents[i] = document
+		amount := float64(document.Transaction.Amount)
+		if amount > max || max == 0 {
+			max = amount
+		}
+		if amount < min || min == 0 {
+			min = amount
+		}
+	}
+
+	minified := p.wc.Minify()
+	for i := range p.documents {
+		document := p.documents[i]
+		amount := float64(document.Transaction.Amount)
+		adjusted := (amount - min) / (max - min)
+		document.AmountScaled = 2*adjusted - 1
+
+		// Calculate a vector based on all the words from all transactions.
+		// TODO Also need to look into Principal Component Analysis
+		vector := make([]float64, len(minified))
+		for word, tfidfValue := range document.TFIDF {
+			index, exists := minified[word]
+			if !exists {
+				continue
+			}
+			vector[index] = tfidfValue
+		}
+		var norm float64
+		for _, value := range vector {
+			norm += value * value
+		}
+		norm = math.Sqrt(norm)
+		for i, value := range vector {
+			vector[i] = value / norm
+		}
```

Comment:
Also if I don't need a ton of precision I might be able to switch to float32 for all of these as well. That would let the SIMD code process 8 items at a time instead of 4.

Reply:
instructions are backwards